notebooks/laplace-mechanism.ipynb

```diff
@@ -0,0 +1,123 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Defining A Custom Privacy Mechanisms\n",
+    "\n",
+    "In this notebook we demonstrate how to extend the accountant to support custom privacy mechanisms.\n",
+    "For the sake of the tutorial we keep things simple by using the well known Laplace mechanism.\n",
+    "\n",
+    "Assume we have a privacy mechanisms with a privacy curve $\\delta(P || Q )$ where $P = \\text{Lap}(\\mu,1) $ and $ Q = \\text{Lap}(0,1)$.\n",
+    "By Proposition B.2 in [[1](https://arxiv.org/pdf/2106.02848.pdf)], the privacy random variables $X$ and $Y$ are given by $ X = | Z | - | Z - \\mu | $ and $ Y = | Z - \\mu | - | Z | $ where $ Z \\sim \\text{Lap}(0,1)$.\n",
+    "\n",
+    "More concretely, the CDF of $Y$ is given by\n",
+    "$$\n",
+    "F_Y(t) =\n",
+    "  \\begin{cases}\n",
+    "    1                                                 & \\quad \\text{if } t > \\mu \\\\\n",
+    "    \\frac{1}{2} \\exp \\left ( \\frac{t-\\mu}{2} \\right ) & \\quad \\text{if } - \\mu \\leq t \\leq \\mu \\\\\n",
+    "    0                                                 & \\quad \\text{otherwise}\n",
+    "  \\end{cases} \\; .\n",
+    "$$"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Implementing the Privacy Random Variable\n",
+    "\n",
+    "Given the derivation of $F_X$, we can now implement the corresponding `LaplaceMechanism` class which is a subclass of `PrivacyRandomVariable`.\n",
+    "The interface is lightweight and only the function `cdf` for the CDF is required to be overridden.\n",
+    "\n",
+    "Optionally, an implementation for the Renyi divergences can be provided (see Definition 3 in [[2](https://ieeexplore.ieee.org/abstract/document/8049725)]).\n",
+    "This allows to automatically compute the required domain size.\n",
+    "If this is not provided the `eps_max` argument in the accountant needs to be set.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "from prv_accountant import PrivacyRandomVariable\n",
+    "\n",
+    "class LaplaceMechanism(PrivacyRandomVariable):\n",
+    "    def __init__(self, mu: float) -> None:\n",
+    "        self.mu = mu\n",
+    "        assert self.mu > 0\n",
+    "\n",
+    "    def cdf(self, t):\n",
+    "        return np.where( t >= self.mu,\n",
+    "            1,\n",
+    "            np.where(np.logical_and(-self.mu < t, t < self.mu),\n",
+    "                1/2*np.exp(1/2*(t-self.mu)),\n",
+    "                0\n",
+    "            )\n",
+    "        )\n",
```

Comment:
NIT, these logic checks are a bit hard to capture when they are nested. In a notebook, where the purpose is for ease of understanding, do you think it would be easier to read with variable names to the `np.where` variables?